File: hermes/app/src/main/java/edu/cmu/cs/dronebrain/interfaces/DroneItf.java
// SPDX-FileCopyrightText: 2023 Carnegie Mellon University - Satyalab
//
// SPDX-License-Identifier: GPL-2.0-only

package edu.cmu.cs.dronebrain.interfaces;

import java.util.ArrayList;
import java.util.Vector;

public interface DroneItf {
    void init() throws Exception;
    void connect() throws Exception;
    void disconnect() throws Exception;
    void takeOff() throws Exception;
    void land() throws Exception;
    void setHome(Double lat, Double lng) throws Exception;
    void moveTo(Double lat, Double lng, Double alt) throws Exception;
    void moveBy(Double x, Double y, Double z) throws Exception; //x,y,z in meters
    void startStreaming(Integer resolution) throws Exception;
    byte[] getVideoFrame() throws Exception;
    void stopStreaming() throws Exception;
    void rotateBy(Double theta) throws Exception;
    void rotateTo(Double theta) throws Exception;
    void setGimbalPose(Double yaw_theta, Double pitch_theta, Double roll_theta) throws Exception;
    void takePhoto() throws Exception;
    void PCMD(Integer pitch, Integer yaw, Integer roll);
    // TODO: Remove this, it's for testing only.
    ArrayList<Double> calculateOffsets(Integer pixel_x, Integer pixel_y, Double leash);
    void trackTarget(Vector<Double> box, Double leash) throws Exception;
    void lookAtTarget(Vector<Double> box) throws Exception;
    void getStatus() throws Exception;
    String getName() throws Exception;
    Double getLat() throws Exception;
    Double getLon() throws Exception;
    Double getAlt() throws Exception;
    Double getExactAlt() throws Exception;
    Double getHeading() throws Exception;
    void hover() throws Exception;
    void cancel() throws Exception;
    void kill() throws Exception;
}

File: droneDSL/postprocess/python/interfaces/DroneItf.py
# SPDX-FileCopyrightText: 2023 Carnegie Mellon University - Satyalab
#
# SPDX-License-Identifier: GPL-2.0-only

from abc import ABC, abstractmethod

class DroneItf(ABC):

    ''' Connection methods '''

    @abstractmethod
    def connect(self):
        pass

    @abstractmethod
    def isConnected(self):
        pass

    @abstractmethod
    def disconnect(self):
        pass
    
    ''' Streaming methods '''

    @abstractmethod
    def startStreaming(self, resolution):
        pass
    
    @abstractmethod
    def getVideoFrame(self):
        pass

    @abstractmethod
    def stopStreaming(self):
        pass

    ''' Take off/ Landing methods '''

    @abstractmethod
    def takeOff(self):
        pass

    @abstractmethod
    def land(self):
        pass

    @abstractmethod
    def setHome(self, lat, lng):
        pass

    @abstractmethod
    def rth(self):
        pass
    
    ''' Movement methods '''

    @abstractmethod
    def PCMD(self, pitch, yaw, roll, gaz, rot):
        pass

    @abstractmethod
    def moveTo(self, lat, lng, alt):
        pass
    
    @abstractmethod
    def moveBy(self, x, y, z, t):
        pass

    @abstractmethod
    def rotateTo(self, theta):
        pass

    @abstractmethod
    def setGimbalPose(self, yaw_theta, pitch_theta, roll_theta):
        pass

    @abstractmethod
    def hover(self):
        pass

    ''' Photography methods ''' 

    @abstractmethod
    def takePhoto(self):
        pass

    @abstractmethod
    def toggleThermal(self, on):
        pass

    ''' Status methods '''

    @abstractmethod
    def getName(self):
        pass

    @abstractmethod
    def getLat(self):
        pass

    @abstractmethod
    def getLng(self):
        pass
    
    @abstractmethod
    def getHeading(self):
        pass

    @abstractmethod
    def getRelAlt(self):
        pass

    @abstractmethod
    def getExactAlt(self):
        pass
    
    @abstractmethod
    def getRSSI(self):
        pass

    @abstractmethod
    def getBatteryPercentage(self):
        pass

    @abstractmethod
    def getMagnetometerReading(self):
        pass

    @abstractmethod
    def getSatellites(self):
        pass

    @abstractmethod
    def getGimbalPitch(self):
        pass
    
    ''' Control methods '''

    @abstractmethod
    def kill(self):
        pass
